I293 no-u-turn sampler

martinjrobins opened the pull request:
Fixes #293 

- Implemented NUTS sampler as a coroutine
- added ability to send a `None` to logging
- added NUTS notebook
- conditionally import NUTS sampler, only implemented for python version > 3.3

@MichaelClerx, when you have time can you have a look at this and see if it's ok. It's mostly complete, I just have to do the conditional importing. In particular, perhaps I should be using the "initial_phase" functionality of the samplers for the dual averaging?
 
@ben: NUTS works great on logistic problem (my problems yesterday were mostly dumb coding mistakes!)

MichaelClerx commented on pints/_logger.py:
```diff
@@ -386,19 +386,22 @@ def log(self, *data):
+                v = next(column)
+                if v is None:
+                    x = ' ' * width
                 else:
```
Could be an elif in case we want to save space :D

martinjrobins replied:
done

MichaelClerx commented on pints/_mcmc/_nuts.py:
```diff
@@ -0,0 +1,557 @@
+
+class NutsState:
+    """
+    Class to hold information about the current state of the NUTS hamiltonian
```
Could have `.` at the end of every sentence?

martinjrobins replied:
done

MichaelClerx commented on pints/_mcmc/_nuts.py:
```diff
@@ -0,0 +1,557 @@
+
+    def update(self, other_state, direction, root):
+        """
+        if root == True, this combines a depth j subtree (`self`) with a depth j+1
```
` ``root == True`` ` etc (double ticks in RST)

martinjrobins replied:
done

MichaelClerx commented on pints/_mcmc/_nuts.py:
```diff
@@ -0,0 +1,557 @@
+        self.s *= int((self.theta_plus - self.theta_minus).dot(self.r_plus) >= 0)
+
+
+@asyncio.coroutine
```
Could we have some comments in here explaining why these need to be coroutines?

martinjrobins replied:
done

MichaelClerx commented on pints/_mcmc/_nuts.py:
```diff
@@ -0,0 +1,557 @@
+
+@asyncio.coroutine
+def leapfrog(theta, L, grad_L, r, epsilon, step_size):
+    """ performs a leapfrog step with step size `epsilon*step_size """
```
`Performs` and ` ``epsilon*step_size``. `

martinjrobins replied:
done

martinjrobins commented on pints/tests/test_mcmc_nuts.py:
```diff
@@ -0,0 +1,150 @@
+
+
+if __name__ == '__main__':
+    print('Add -v for more debug output')
```
done

martinjrobins commented on pints/tests/test_mcmc_nuts.py:
```diff
@@ -0,0 +1,150 @@
+
+from shared import StreamCapture
+
+debug = False
```
done

martinjrobins commented on pints/tests/test_mcmc_nuts.py:
```diff
@@ -0,0 +1,150 @@
+# Tests the basic methods of the Hamiltonian MCMC routine.
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017-2019, University of Oxford.
```
done

ben18785 commented on examples/README.md:
```diff
@@ -51,6 +51,7 @@ relevant code.
 
 ### MCMC with gradients
 - [Hamiltonian MCMC](./sampling-hamiltonian-mcmc.ipynb)
+- [No-U-Turn MCMC](./sampling-nuts-mcmc.ipynb)
```
Very minor but could we keep this alphabetical.

martinjrobins replied:
done

martinjrobins commented:
thanks @ben18785, we don't track divergent iterations for nuts, but can add this in a similar way to hamiltonianmcmc. I'll alter the notebook as you suggest as well

ben18785 commented:
Cool -- yep, would be good to track these I think. Ta!

On Wed, Feb 5, 2020 at 9:19 AM Martin Robinson <notifications@github.com>
wrote:

> thanks @ben18785 <https://github.com/ben18785>, we don't track divergent
> iterations for nuts, but can add this in a similar way to hamiltonianmcmc.
> I'll alter the notebook as you suggest as well
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pints-team/pints/pull/1031?email_source=notifications&email_token=ABCILKHU2YTHT4GP5SU3QHTRBKACVA5CNFSM4KNC4FM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEK2WRHI#issuecomment-582314141>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCILKEFH7GXYUHVSB7JDHTRBKACVANCNFSM4KNC4FMQ>
> .
>

ben18785 commented on pints/_mcmc/_nuts.py:
```diff
@@ -0,0 +1,571 @@
+
+        # build up an integration path with 2^j points, stopping when we either
+        # encounter a U-Turn, or reach a max number of points 2^10
+        while j < 10 and state.s == 1:
```
@martinjrobins Whilst I remember, can we make the tree depth a hyper-parameter?

martinjrobins replied:
done

MichaelClerx commented:
Tantalisingly close to being mergeable :D !

martinjrobins commented:
Almost! I'm just adding the multinomial sampling that they use in Stan

MichaelClerx commented:
Cool! I just added it to the docs and fixed the test skipping. Hope that doesn't interfere

ben18785 commented on pints/_mcmc/_nuts.py:
```diff
@@ -0,0 +1,719 @@
+    r"""
+
+    Implements No U-Turn Sampler (NUTS) with dual averaging, as described in
+    Algorithm 6 in [1]_.
```
Cool. Can we add a note to say that it either does Algo 6 in [1] or does multinomial sampling. In the latter case, can we include pseudocode or something (seeing as it's not written anywhere in Stan!).

martinjrobins replied:
where should the pseudocode go? In the docstring, can this be formatted nicely?

martinjrobins commented:
So I've implemented the multinomial sampling, dense or diagonal mass matrix, and full adaption of epsilon and mass matrix. The multinomial sampling seems to be slightly worse than the slice sampling, so I'm thinking perhaps I've implemented it wrong.....

Note for future: I have split off the adaption of epsilon and mass matrix to a separate class, so this can be used in HMC as well later on

martinjrobins commented:
closing this as development of nuts is continuing on `nuts-comparison` branch

MichaelClerx commented:
Hey @martinjrobins ! Is it OK to delete this branch now?

MichaelClerx commented:
0 new commits compared to master, I see, so closing!
